Firstly I load the packages I need.

In [34]:
import io
import requests
import pandas as pd
import numpy as np

Secondly I use the pandas .read_html method to scrape the table with the neighbourhoods in Toronto from the Wikipedia page. I specified by using [0] for the particular table I want since the page contains more than one table. 

In [20]:
Toronto = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M', header=0,
                     )
Toronto[0]

Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A      Queen's Park   
8        M8A      Not assigned   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
12       M2B      Not assigned   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
19       M7B      Not assigned   
20       M8B      Not assigned   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M2C      Not assigned   
..       ...               ...   
257      M9X      Not assigned   
258      M1Y      Not assigned   
259      M2Y      Not assigned   
260      M3Y      Not assigned   
261      M4Y  Downtown Toronto   
262      M5Y      Not assigned   
263      M6Y      Not assigned   
264      M7Y      East Toronto   
265      M8Y         Etobicoke   
266      M8Y         Etobicoke   
267      M8Y         Etobicoke   
268      M8Y         Etobicoke   
269      M8Y         Etobicoke   
270      M8Y         Etobicoke   
271      M8Y         Etobicoke   
272      M8Y         Etobicoke   
273      M9Y      Not assigned   
274      M1Z      Not assigned   
275      M2Z      Not assigned   
276      M3Z      Not assigned   
277      M4Z      Not assigned   
278      M5Z      Not assigned   
279      M6Z      Not assigned   
280      M7Z      Not assigned   
281      M8Z         Etobicoke   
282      M8Z         Etobicoke   
283      M8Z         Etobicoke   
284      M8Z         Etobicoke   
285      M8Z         Etobicoke   
286      M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                                         Harbourfront  
5                                     Lawrence Heights  
6                                       Lawrence Manor  
7                                         Not assigned  
8                                         Not assigned  
9                                         Queen's Park  
10                                               Rouge  
11                                             Malvern  
12                                        Not assigned  
13                                     Don Mills North  
14                                    Woodbine Gardens  
15                                       Parkview Hill  
16                                             Ryerson  
17                                     Garden District  
18                                           Glencairn  
19                                        Not assigned  
20                                        Not assigned  
21                                          Cloverdale  
22                                           Islington  
23                                        Martin Grove  
24                                    Princess Gardens  
25                                     West Deane Park  
26                                      Highland Creek  
27                                          Rouge Hill  
28                                          Port Union  
29                                        Not assigned  
..                                                 ...  
257                                       Not assigned  
258           

The scraped table contains 287 rows and 3 columns. Then I convert the table into a pandas dataframe called df.

In [21]:
df = pd.DataFrame(Toronto[0])
df

Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A      Queen's Park   
8        M8A      Not assigned   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
12       M2B      Not assigned   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
19       M7B      Not assigned   
20       M8B      Not assigned   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M2C      Not assigned   
..       ...               ...   
257      M9X      Not assigned   
258      M1Y      Not assigned   
259      M2Y      Not assigned   
260      M3Y      Not assigned   
261      M4Y  Downtown Toronto   
262      M5Y      Not assigned   
263      M6Y      Not assigned   
264      M7Y      East Toronto   
265      M8Y         Etobicoke   
266      M8Y         Etobicoke   
267      M8Y         Etobicoke   
268      M8Y         Etobicoke   
269      M8Y         Etobicoke   
270      M8Y         Etobicoke   
271      M8Y         Etobicoke   
272      M8Y         Etobicoke   
273      M9Y      Not assigned   
274      M1Z      Not assigned   
275      M2Z      Not assigned   
276      M3Z      Not assigned   
277      M4Z      Not assigned   
278      M5Z      Not assigned   
279      M6Z      Not assigned   
280      M7Z      Not assigned   
281      M8Z         Etobicoke   
282      M8Z         Etobicoke   
283      M8Z         Etobicoke   
284      M8Z         Etobicoke   
285      M8Z         Etobicoke   
286      M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                                         Harbourfront  
5                                     Lawrence Heights  
6                                       Lawrence Manor  
7                                         Not assigned  
8                                         Not assigned  
9                                         Queen's Park  
10                                               Rouge  
11                                             Malvern  
12                                        Not assigned  
13                                     Don Mills North  
14                                    Woodbine Gardens  
15                                       Parkview Hill  
16                                             Ryerson  
17                                     Garden District  
18                                           Glencairn  
19                                        Not assigned  
20                                        Not assigned  
21                                          Cloverdale  
22                                           Islington  
23                                        Martin Grove  
24                                    Princess Gardens  
25                                     West Deane Park  
26                                      Highland Creek  
27                                          Rouge Hill  
28                                          Port Union  
29                                        Not assigned  
..                                                 ...  
257                                       Not assigned  
258           

I then delete all cells with a Borough that is Not assigned so that I only process the cells that have an assigned borough.

In [22]:
df = df[df.Borough != 'Not assigned']
df

Postcode           Borough  \
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A      Queen's Park   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
33       M5C  Downtown Toronto   
34       M6C              York   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
..       ...               ...   
234      M9V         Etobicoke   
235      M1W       Scarborough   
238      M4W  Downtown Toronto   
239      M5W  Downtown Toronto   
242      M8W         Etobicoke   
243      M8W         Etobicoke   
244      M9W         Etobicoke   
245      M1X       Scarborough   
248      M4X  Downtown Toronto   
249      M4X  Downtown Toronto   
250      M5X  Downtown Toronto   
251      M5X  Downtown Toronto   
254      M8X         Etobicoke   
255      M8X         Etobicoke   
256      M8X         Etobicoke   
261      M4Y  Downtown Toronto   
264      M7Y      East Toronto   
265      M8Y         Etobicoke   
266      M8Y         Etobicoke   
267      M8Y         Etobicoke   
268      M8Y         Etobicoke   
269      M8Y         Etobicoke   
270      M8Y         Etobicoke   
271      M8Y         Etobicoke   
272      M8Y         Etobicoke   
281      M8Z         Etobicoke   
282      M8Z         Etobicoke   
283      M8Z         Etobicoke   
284      M8Z         Etobicoke   
285      M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                                         Harbourfront  
5                                     Lawrence Heights  
6                                       Lawrence Manor  
7                                         Not assigned  
9                                         Queen's Park  
10                                               Rouge  
11                                             Malvern  
13                                     Don Mills North  
14                                    Woodbine Gardens  
15                                       Parkview Hill  
16                                             Ryerson  
17                                     Garden District  
18                                           Glencairn  
21                                          Cloverdale  
22                                           Islington  
23                                        Martin Grove  
24                                    Princess Gardens  
25                                     West Deane Park  
26                                      Highland Creek  
27                                          Rouge Hill  
28                                          Port Union  
30                                     Flemingdon Park  
31                                     Don Mills South  
32                                    Woodbine Heights  
33                                      St. James Town  
34                                  Humewood-Cedarvale  
37                                   Bloordale Gardens  
38                                            Eringate  
..                                                 ...  
234                                        Thistletown  
235           

This reduces my dataframe from 287 rows to 210 rows, meaning that 77 cells had a Borough that was Not assigned. 

I then combine into one row all Neighbourhoods with the same Postcode, with the Neighbourhoods separated with a comma. To do that, firstly I use the .groupby method on condition of the Neighbourhoods having the same Postcode and then use the .apply method to execute the grouping and the .join method to separate the Neighbourhoods with a coma and save the result in a new dataframe called df1.

In [23]:
df1 = df.groupby('Postcode')['Neighbourhood'].apply(','.join).reset_index()
df1

Postcode                                      Neighbourhood
0        M1B                                      Rouge,Malvern
1        M1C               Highland Creek,Rouge Hill,Port Union
2        M1E                    Guildwood,Morningside,West Hill
3        M1G                                             Woburn
4        M1H                                          Cedarbrae
5        M1J                                Scarborough Village
6        M1K          East Birchmount Park,Ionview,Kennedy Park
7        M1L                      Clairlea,Golden Mile,Oakridge
8        M1M      Cliffcrest,Cliffside,Scarborough Village West
9        M1N                         Birch Cliff,Cliffside West
10       M1P  Dorset Park,Scarborough Town Centre,Wexford He...
11       M1R                                   Maryvale,Wexford
12       M1S                                          Agincourt
13       M1T              Clarks Corners,Sullivan,Tam O'Shanter
14       M1V  Agincourt North,L'Amoreaux East,Milliken,Steel...
15       M1W                                    L'Amoreaux West
16       M1X                                        Upper Rouge
17       M2H                                  Hillcrest Village
18       M2J                         Fairview,Henry Farm,Oriole
19       M2K                                    Bayview Village
20       M2L                            Silver Hills,York Mills
21       M2M                             Newtonbrook,Willowdale
22       M2N                                   Willowdale South
23       M2P                                    York Mills West
24       M2R                                    Willowdale West
25       M3A                                          Parkwoods
26       M3B                                    Don Mills North
27       M3C                    Flemingdon Park,Don Mills South
28       M3H      Bathurst Manor,Downsview North,Wilson Heights
29       M3J                     Northwood Park,York University
..       ...                                                ...
73       M6C                                 Humewood-Cedarvale
74       M6E                                Caledonia-Fairbanks
75       M6G                                           Christie
76       M6H                        Dovercourt Village,Dufferin
77       M6J                            Little Portugal,Trinity
78       M6K         Brockton,Exhibition Place,Parkdale Village
79       M6L                   Downsview,North Park,Upwood Park
80       M6M        Del Ray,Keelesdale,Mount Dennis,Silverthorn
81       M6N                       The Junction North,Runnymede
82       M6P                       High Park,The Junction South
83       M6R                              Parkdale,Roncesvalles
84       M6S                                  Runnymede,Swansea
85       M7A                                       Not assigned
86       M7R              Canada Post Gateway Processing Centre
87       M7Y  Business Reply Mail Processing Centre 969 Eastern
88       M8V         Humber Bay Shores,Mimico South,New Toronto
89       M8W                              Alderwood,Long Branch
90       M8X        The Kingsway,Montgomery Road,Old Mill North
91       M8Y  Humber Bay,King's Mill Park,Kingsway Park Sout...
92       M8Z  Kingsway Park South West,Mimico NW,The Queensw...
93       M9A                                       Queen's Park
94       M9B  Cloverdale,Islington,Martin Grove,Princess Gar...
95       M9C  Bloordale Gardens,Eringate,Markland Wood,Old B...
96       M9L                                      Humber Summit
97       M9M                                    Emery,Humberlea
98       M9N                                             Weston
99       M9P                                          Westmount
100      M9R  Kingsview Village,Martin Grove Gardens,Richvie...
101      M9V  Albion Gardens,Beaumond Heights,Humbergate,Jam...
102      M9W                                          Northwest

[103 rows x 2 columns]

The returned output above excludes the Borough column. I use the pandas .merge method to combine the Borough column from the df dataframe and the df1 dataframe on condition that they're merged on matching the Borough to the Neighbourhood based on the Postcode and I save the results in a new dataframe called df2.

In [24]:
df2 = pd.merge(df1, df[["Postcode", "Borough"]], on="Postcode")
df2

Postcode                                      Neighbourhood       Borough
0        M1B                                      Rouge,Malvern   Scarborough
1        M1B                                      Rouge,Malvern   Scarborough
2        M1C               Highland Creek,Rouge Hill,Port Union   Scarborough
3        M1C               Highland Creek,Rouge Hill,Port Union   Scarborough
4        M1C               Highland Creek,Rouge Hill,Port Union   Scarborough
5        M1E                    Guildwood,Morningside,West Hill   Scarborough
6        M1E                    Guildwood,Morningside,West Hill   Scarborough
7        M1E                    Guildwood,Morningside,West Hill   Scarborough
8        M1G                                             Woburn   Scarborough
9        M1H                                          Cedarbrae   Scarborough
10       M1J                                Scarborough Village   Scarborough
11       M1K          East Birchmount Park,Ionview,Kennedy Park   Scarborough
12       M1K          East Birchmount Park,Ionview,Kennedy Park   Scarborough
13       M1K          East Birchmount Park,Ionview,Kennedy Park   Scarborough
14       M1L                      Clairlea,Golden Mile,Oakridge   Scarborough
15       M1L                      Clairlea,Golden Mile,Oakridge   Scarborough
16       M1L                      Clairlea,Golden Mile,Oakridge   Scarborough
17       M1M      Cliffcrest,Cliffside,Scarborough Village West   Scarborough
18       M1M      Cliffcrest,Cliffside,Scarborough Village West   Scarborough
19       M1M      Cliffcrest,Cliffside,Scarborough Village West   Scarborough
20       M1N                         Birch Cliff,Cliffside West   Scarborough
21       M1N                         Birch Cliff,Cliffside West   Scarborough
22       M1P  Dorset Park,Scarborough Town Centre,Wexford He...   Scarborough
23       M1P  Dorset Park,Scarborough Town Centre,Wexford He...   Scarborough
24       M1P  Dorset Park,Scarborough Town Centre,Wexford He...   Scarborough
25       M1R                                   Maryvale,Wexford   Scarborough
26       M1R                                   Maryvale,Wexford   Scarborough
27       M1S                                          Agincourt   Scarborough
28       M1T              Clarks Corners,Sullivan,Tam O'Shanter   Scarborough
29       M1T              Clarks Corners,Sullivan,Tam O'Shanter   Scarborough
..       ...                                                ...           ...
180      M8Z  Kingsway Park South West,Mimico NW,The Queensw...     Etobicoke
181      M8Z  Kingsway Park South West,Mimico NW,The Queensw...     Etobicoke
182      M9A                                       Queen's Park  Queen's Park
183      M9B  Cloverdale,Islington,Martin Grove,Princess Gar...     Etobicoke
184      M9B  Cloverdale,Islington,Martin Grove,Princess Gar...     Etobicoke
185      M9B  Cloverdale,Islington,Martin Grove,Princess Gar...     Etobicoke
186      M9B  Cloverdale,Islington,Martin Grove,Princess Gar...     Etobicoke
187      M9B  Cloverdale,Islington,Martin Grove,Princess Gar...     Etobicoke
188      M9C  Bloordale Gardens,Eringate,Markland Wood,Old B...     Etobicoke
189      M9C  Bloordale Gardens,Eringate,Markland Wood,Old B...     Etobicoke
190      M9C  Bloordale Gardens,Eringate,Markland Wood,Old B...     Etobicoke
191      M9C  Bloordale Gardens,Eringate,Markland Wood,Old B...     Etobicoke
192      M9L                                      Humber Summit    North York
193      M9M                                    Emery,Humberlea    North York
194      M9M                                    Emery,Humberlea    North York
195      M9N                                             Weston          York
196      M9P                                          Westmount     Etobicoke
197      M9R  Kingsview Village,Martin Grove Gardens,Richvie...     Etobicoke
198      M9R  Kingsview Village,Martin Grove Gardens,Richvie...     Etobicoke
199      M9R  Kingsview Vi

I then rearrange the order of my columns to start with Postcode, followed by Borough and thirdly Neighbourhood.

In [25]:
df2 = df2[['Postcode', 'Borough', 'Neighbourhood']]
df2

Postcode       Borough                                      Neighbourhood
0        M1B   Scarborough                                      Rouge,Malvern
1        M1B   Scarborough                                      Rouge,Malvern
2        M1C   Scarborough               Highland Creek,Rouge Hill,Port Union
3        M1C   Scarborough               Highland Creek,Rouge Hill,Port Union
4        M1C   Scarborough               Highland Creek,Rouge Hill,Port Union
5        M1E   Scarborough                    Guildwood,Morningside,West Hill
6        M1E   Scarborough                    Guildwood,Morningside,West Hill
7        M1E   Scarborough                    Guildwood,Morningside,West Hill
8        M1G   Scarborough                                             Woburn
9        M1H   Scarborough                                          Cedarbrae
10       M1J   Scarborough                                Scarborough Village
11       M1K   Scarborough          East Birchmount Park,Ionview,Kennedy Park
12       M1K   Scarborough          East Birchmount Park,Ionview,Kennedy Park
13       M1K   Scarborough          East Birchmount Park,Ionview,Kennedy Park
14       M1L   Scarborough                      Clairlea,Golden Mile,Oakridge
15       M1L   Scarborough                      Clairlea,Golden Mile,Oakridge
16       M1L   Scarborough                      Clairlea,Golden Mile,Oakridge
17       M1M   Scarborough      Cliffcrest,Cliffside,Scarborough Village West
18       M1M   Scarborough      Cliffcrest,Cliffside,Scarborough Village West
19       M1M   Scarborough      Cliffcrest,Cliffside,Scarborough Village West
20       M1N   Scarborough                         Birch Cliff,Cliffside West
21       M1N   Scarborough                         Birch Cliff,Cliffside West
22       M1P   Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...
23       M1P   Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...
24       M1P   Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...
25       M1R   Scarborough                                   Maryvale,Wexford
26       M1R   Scarborough                                   Maryvale,Wexford
27       M1S   Scarborough                                          Agincourt
28       M1T   Scarborough              Clarks Corners,Sullivan,Tam O'Shanter
29       M1T   Scarborough              Clarks Corners,Sullivan,Tam O'Shanter
..       ...           ...                                                ...
180      M8Z     Etobicoke  Kingsway Park South West,Mimico NW,The Queensw...
181      M8Z     Etobicoke  Kingsway Park South West,Mimico NW,The Queensw...
182      M9A  Queen's Park                                       Queen's Park
183      M9B     Etobicoke  Cloverdale,Islington,Martin Grove,Princess Gar...
184      M9B     Etobicoke  Cloverdale,Islington,Martin Grove,Princess Gar...
185      M9B     Etobicoke  Cloverdale,Islington,Martin Grove,Princess Gar...
186      M9B     Etobicoke  Cloverdale,Islington,Martin Grove,Princess Gar...
187      M9B     Etobicoke  Cloverdale,Islington,Martin Grove,Princess Gar...
188      M9C     Etobicoke  Bloordale Gardens,Eringate,Markland Wood,Old B...
189      M9C     Etobicoke  Bloordale Gardens,Eringate,Markland Wood,Old B...
190      M9C     Etobicoke  Bloordale Gardens,Eringate,Markland Wood,Old B...
191      M9C     Etobicoke  Bloordale Gardens,Eringate,Markland Wood,Old B...
192      M9L    North York                                      Humber Summit
193      M9M    North York                                    Emery,Humberlea
194      M9M    North York                                    Emery,Humberlea
195      M9N          York                                             Weston
196      M9P     Etobicoke                                          Westmount
197      M9R     Etobicoke  Kingsview Village,Martin Grove Gardens,Richvie...
198      M9R     Etobicoke  Kingsview Village,Martin Grove Gardens,Richvie...
199      M9R     Etobicoke

I then drop all resulting duplicates based on the same Postcode and save the results in a new dataframe called df3.

In [26]:
df3 = df2.drop_duplicates(subset="Postcode")
df3

Postcode           Borough  \
0        M1B       Scarborough   
2        M1C       Scarborough   
5        M1E       Scarborough   
8        M1G       Scarborough   
9        M1H       Scarborough   
10       M1J       Scarborough   
11       M1K       Scarborough   
14       M1L       Scarborough   
17       M1M       Scarborough   
20       M1N       Scarborough   
22       M1P       Scarborough   
25       M1R       Scarborough   
27       M1S       Scarborough   
28       M1T       Scarborough   
31       M1V       Scarborough   
35       M1W       Scarborough   
36       M1X       Scarborough   
37       M2H        North York   
38       M2J        North York   
41       M2K        North York   
42       M2L        North York   
44       M2M        North York   
46       M2N        North York   
47       M2P        North York   
48       M2R        North York   
49       M3A        North York   
50       M3B        North York   
51       M3C        North York   
53       M3H        North York   
56       M3J        North York   
..       ...               ...   
133      M6C              York   
134      M6E              York   
135      M6G  Downtown Toronto   
136      M6H      West Toronto   
138      M6J      West Toronto   
140      M6K      West Toronto   
143      M6L        North York   
146      M6M              York   
150      M6N              York   
152      M6P      West Toronto   
154      M6R      West Toronto   
156      M6S      West Toronto   
158      M7A      Queen's Park   
159      M7R       Mississauga   
160      M7Y      East Toronto   
161      M8V         Etobicoke   
164      M8W         Etobicoke   
166      M8X         Etobicoke   
169      M8Y         Etobicoke   
177      M8Z         Etobicoke   
182      M9A      Queen's Park   
183      M9B         Etobicoke   
188      M9C         Etobicoke   
192      M9L        North York   
193      M9M        North York   
195      M9N              York   
196      M9P         Etobicoke   
197      M9R         Etobicoke   
201      M9V         Etobicoke   
209      M9W         Etobicoke   

                                         Neighbourhood  
0                                        Rouge,Malvern  
2                 Highland Creek,Rouge Hill,Port Union  
5                      Guildwood,Morningside,West Hill  
8                                               Woburn  
9                                            Cedarbrae  
10                                 Scarborough Village  
11           East Birchmount Park,Ionview,Kennedy Park  
14                       Clairlea,Golden Mile,Oakridge  
17       Cliffcrest,Cliffside,Scarborough Village West  
20                          Birch Cliff,Cliffside West  
22   Dorset Park,Scarborough Town Centre,Wexford He...  
25                                    Maryvale,Wexford  
27                                           Agincourt  
28               Clarks Corners,Sullivan,Tam O'Shanter  
31   Agincourt North,L'Amoreaux East,Milliken,Steel...  
35                                     L'Amoreaux West  
36                                         Upper Rouge  
37                                   Hillcrest Village  
38                          Fairview,Henry Farm,Oriole  
41                                     Bayview Village  
42                             Silver Hills,York Mills  
44                              Newtonbrook,Willowdale  
46                                    Willowdale South  
47                                     York Mills West  
48                                     Willowdale West  
49                                           Parkwoods  
50                                     Don Mills North  
51                     Flemingdon Park,Don Mills South  
53       Bathurst Manor,Downsview North,Wilson Heights  
56                      Northwood Park,York University  
..                                                 ...  
133                                 Humewood-Cedarvale  
134           

The code below is to find out how many Neighbourhoods are Not assigned. The result is just one, with a Borough named Queen's Park.

In [27]:
print(df3.loc[df3['Neighbourhood'] == 'Not assigned'])

    Postcode       Borough Neighbourhood
158      M7A  Queen's Park  Not assigned


I then replace the Not assigned Neighbourhood with its Borough name by using the .loc method.

In [28]:
df3.loc[df3['Neighbourhood'] == 'Not assigned'] = "Queen's Park"
df3

/usr/local/lib/python3.5/dist-packages/pandas/core/indexing.py:189: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)
/usr/local/lib/python3.5/dist-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


Postcode           Borough  \
0             M1B       Scarborough   
2             M1C       Scarborough   
5             M1E       Scarborough   
8             M1G       Scarborough   
9             M1H       Scarborough   
10            M1J       Scarborough   
11            M1K       Scarborough   
14            M1L       Scarborough   
17            M1M       Scarborough   
20            M1N       Scarborough   
22            M1P       Scarborough   
25            M1R       Scarborough   
27            M1S       Scarborough   
28            M1T       Scarborough   
31            M1V       Scarborough   
35            M1W       Scarborough   
36            M1X       Scarborough   
37            M2H        North York   
38            M2J        North York   
41            M2K        North York   
42            M2L        North York   
44            M2M        North York   
46            M2N        North York   
47            M2P        North York   
48            M2R        North York   
49            M3A        North York   
50            M3B        North York   
51            M3C        North York   
53            M3H        North York   
56            M3J        North York   
..            ...               ...   
133           M6C              York   
134           M6E              York   
135           M6G  Downtown Toronto   
136           M6H      West Toronto   
138           M6J      West Toronto   
140           M6K      West Toronto   
143           M6L        North York   
146           M6M              York   
150           M6N              York   
152           M6P      West Toronto   
154           M6R      West Toronto   
156           M6S      West Toronto   
158  Queen's Park      Queen's Park   
159           M7R       Mississauga   
160           M7Y      East Toronto   
161           M8V         Etobicoke   
164           M8W         Etobicoke   
166           M8X         Etobicoke   
169           M8Y         Etobicoke   
177           M8Z         Etobicoke   
182           M9A      Queen's Park   
183           M9B         Etobicoke   
188           M9C         Etobicoke   
192           M9L        North York   
193           M9M        North York   
195           M9N              York   
196           M9P         Etobicoke   
197           M9R         Etobicoke   
201           M9V         Etobicoke   
209           M9W         Etobicoke   

                                         Neighbourhood  
0                                        Rouge,Malvern  
2                 Highland Creek,Rouge Hill,Port Union  
5                      Guildwood,Morningside,West Hill  
8                                               Woburn  
9                                            Cedarbrae  
10                                 Scarborough Village  
11           East Birchmount Park,Ionview,Kennedy Park  
14                       Clairlea,Golden Mile,Oakridge  
17       Cliffcrest,Cliffside,Scarborough Village West  
20                          Birch Cliff,Cliffside West  
22   Dorset Park,Scarborough Town Centre,Wexford He...  
25                                    Maryvale,Wexford  
27                                           Agincourt  
28               Clarks Corners,Sullivan,Tam O'Shanter  
31   Agincourt North,L'Amoreaux East,Milliken,Steel...  
35                                     L'Amoreaux West  
36                                         Upper Rouge  
37                                   Hillcrest Village  
38                          Fairview,Henry Farm,Oriole  
41                                     Bayview Village  
42                             Silver Hills,York Mills  
44                              Newtonbrook,Willowdale  
46                                    Willowdale South  
47                                     York Mills West  
48                                     Willowdale West  
49                                           Parkwoods  
50                                     Don Mills No

To ascertain the number of rows and columns in my final df3 dataframe, I use the .shape method and this shows that my final dataframe has 103 rows and 3 columns.

In [29]:
df4.shape

(103, 3)

First I use the pandas .read_csv method to read the cvs file containing the Latitudes and Longitudes of the Toronto Neighbourhoods based on Postcodes and I save the csv data in a pandas dataframe called Coords.

In [41]:
Coords = pd.read_csv("https://cocl.us/Geospatial_data") 
Coords

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
..          ...        ...        ...
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

I then rename the Postal Code column in the Coords dataframe to Postcode like the df3 dataframe to make it easier for me to merge the 2 dataframes together.

In [42]:
Coords.rename(columns={'Postal Code':'Postcode'}, inplace=True)

print(Coords.columns)

Index(['Postcode', 'Latitude', 'Longitude'], dtype='object')


Using the pandas .merge method to combine the two dataframes (df3 and Coords) together on condition that they're merged on the Postcode being the same and I save the results in a new dataframe called df4.

In [43]:
df4 = pd.merge(df3, Coords[["Postcode", "Latitude", "Longitude"]], on="Postcode")
df4

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
..       ...               ...   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7R       Mississauga   
86       M7Y      East Toronto   
87       M8V         Etobicoke   
88       M8W         Etobicoke   
89       M8X         Etobicoke   
90       M8Y         Etobicoke   
91       M8Z         Etobicoke   
92       M9A      Queen's Park   
93       M9B         Etobicoke   
94       M9C         Etobicoke   
95       M9L        North York   
96       M9M        North York   
97       M9N              York   
98       M9P         Etobicoke   
99       M9R         Etobicoke   
100      M9V         Etobicoke   
101      M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                        Rouge,Malvern  43.806686 -79.194353  
1                 Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
2                      Guildwood,Morningside,West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
5                                  Scarborough Village  43.744734 -79.239476  
6            East Birchmount Park,Ionview,Kennedy Park  43.727929 -79.262029  
7                        Clairlea,Golden Mile,Oakridge  43.711112 -79.284577  
8        Cliffcrest,Cliffside,Scarborough Village West  43.716316 -79.239476  
9                           Birch Cliff,Cliffside West  43.692657 -79.264848  
10   Dorset Park,Scarborough Town Centre,Wexford He...  43.757410 -79.273304  
11                                    Maryvale,Wexford  43.750072 -79.295849  
12                                           Agincourt  43.794200 -79.262029  
13               Clarks Corners,Sullivan,Tam O'Shanter  43.781638 -79.304302  
14   Agincourt North,L'Amoreaux East,Milliken,Steel...  43.815252 -79.284577  
15                                     L'Amoreaux West  43.799525 -79.318389  
16                                         Upper Rouge  43.836125 -79.205636  
17                                   Hillcrest Village  43.803762 -79.363452  
18                          Fairview,Henry Farm,Oriole  43.778517 -79.346556  
19                                     Bayview Village  43.786947 -79.385975  
20                             Silver Hills,York Mills  43.757490 -79.374714  
21                              Newtonbrook,Willowdale  43.789053 -79.408493  
22                                    Willowdale South  43.770120 -79.408493  

The resulting dataframe after the merge has one less row, meaning that one row was dropped maybe due to a mismatch.